SDIOBlockDevice.cpp

```diff
@@ -0,0 +1,378 @@
+    {
+        unlock();
+        return BD_ERROR_OK;
+    }
+
+    if (isPresent() == false)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_DEVICE;
+    }
+
+    int status = SD_Init();
+    if (BD_ERROR_OK != status)
+    {
+        unlock();
+        return BD_ERROR_DEVICE_ERROR;
+    }
+
+    SD_GetCardInfo(&_cardInfo);
+    _is_initialized = true;
+    debug_if(SD_DBG, "SD initialized: type: %d  version: %d  class: %d\n",
+             _cardInfo.CardType, _cardInfo.CardVersion, _cardInfo.Class);
+    debug_if(SD_DBG, "SD size: %d MB\n",
+             _cardInfo.LogBlockNbr / 2 / 1024);
+
+    // get sectors count from cardinfo
+    _sectors = _cardInfo.LogBlockNbr;
+    if (BLOCK_SIZE_HC != _cardInfo.BlockSize)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_UNSUPPORTED_BLOCKSIZE;
+    }
+
+    unlock();
+    return status;
+}
+
+int SDIOBlockDevice::deinit()
+{
+    debug_if(SD_DBG, "deinit Card...\r\n");
+    lock();
+
+    if (!_is_initialized)
+    {
+        _init_ref_count = 0;
+        unlock();
+        return BD_ERROR_OK;
+    }
+
+    _init_ref_count--;
+
+    if (_init_ref_count)
+    {
+        unlock();
+        return BD_ERROR_OK;
+    }
+
+    int status = SD_DeInit();
+    _is_initialized = false;
+
+    _sectors = 0;
+
+    unlock();
+    return status;
+}
+
+int SDIOBlockDevice::read(void *b, bd_addr_t addr, bd_size_t size)
+{
+    //debug_if(SD_DBG, "read Card...\r\n");
+    lock();
+    if (isPresent() == false)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_DEVICE;
+    }
+    if (!is_valid_read(addr, size))
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_PARAMETER;
+    }
+
+    if (!_is_initialized)
+    {
+        unlock();
+        return SD_BLOCK_DEVICE_ERROR_NO_INIT;
+    }
+
+    uint32_t *buffer = static_cast<uint32_t *>(b);
+
+    // ReadBlocks uses byte unit address
+    // SDHC and SDXC Cards different addressing is handled in ReadBlocks()
+    bd_addr_t blockCnt = size / _block_size;
+    addr = addr / _block_size;
+
+    // make sure card is ready
+    while (SD_GetCardState() != SD_TRANSFER_OK)
+    {
+        // wait until SD ready
+        wait_ms(1);
+    }
+
+    // receive the data : one block/ multiple blocks is handled in ReadBlocks()
+    int status = SD_ReadBlocks_DMA(buffer, addr, blockCnt);
+    debug_if(SD_DBG, "ReadBlocks dbgtest addr: %lld  blockCnt: %lld \n", addr, blockCnt);
+
+    if (status == MSD_OK)
+    {
+        // wait until DMA finished
+        while (SD_DMA_ReadPending() != SD_TRANSFER_OK)
+        {
+            uint32_t tickstart = HAL_GetTick();
+            if ((HAL_GetTick() - tickstart) >= MBED_CONF_SD_TIMEOUT)
+            {
+                unlock();
+                return SD_BLOCK_DEVICE_ERROR_READBLOCKS;
+            }
+        }
+        // make sure card is ready
+        while (SD_GetCardState() != SD_TRANSFER_OK)
+        {
+            // wait until SD ready
+            wait_ms(10);
```

Comment:
why is it wait 10 msec here, and 1 msec in other places?

Reply:
I'm testing shorter timeouts now and I have added timeout checks to all loops. In this case, the card should be ready and the check is too make sure that a previous operation is not blocking.
The timeout check with high frequent HAL_Tick() calls may cause problems because there are many calls to locking/unlocking for the ticker access.